keps/sig-api-machinery/5000-api-linting-crd-schema-tooling/README.md

```diff
@@ -0,0 +1,522 @@
+
+## Motivation
+
+Aside from existing documentation on [Kubernetes API conventions][kube-api-conventions] and [making changes to Kubernetes APIs][making-changes-to-apis],
+there is little to no tooling that enable developers to ensure they are following best practices when developing Kubernetes-native APIs.
+
+This KEP aims to improve the Kubernetes-native API development experience by adding two sig-api-machinery subprojects for:
+
+- Linting Kubernetes APIs written in Go based on the [Kubernetes API conventions][kube-api-conventions]
+- Validating changes to CustomResourceDefinition schemas in YAML
+
+[kube-api-conventions]:(https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+[making-changes-to-apis]:(https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+
+### Goals
+
+- Create tooling to help developers writing Kubernetes APIs in Go to follow best practices and Kubernetes API conventions
+- Create tooling to help developers writing CustomResourceDefinitions avoid making breaking changes
+
+### Non-Goals
+
+- Create tooling for validating CustomResourceDefinition schema changes at admission time
+    - This is a non-goal of this KEP in particular to keep the scope of this KEP to only include the individual tooling. A separate KEP may be written, focusing on addressing this goal.
+
+## Proposal
+
+This KEP introduces two new sig-api-machinery subprojects.
+The projects are intended to complement each other, and take lessons learned from [Kubernetes API conventions][kube-api-conventions] and our experience
+with [making changes to Kubernetes APIs][making-changes-to-apis], to enable the wider Kubernetes ecosystem to follow best practices and avoid breaking changes.
+
+### Go-based Kubernetes API Linting
+
+A linter, built as a `golangci-lint` plugin, that evaluates Go type defintions (typically `_types.go`), and flags deviations from Kubernetes API conventions and best practices.
+
+The linter will be configurable, allowing users to choose which of the rules they wish to adhere to, and which they wish to ignore.
+
+The linter will focus, as much as possible on providing a "built-ins" experience by default, targetting rules for types in Kubernetes/Kubernetes.
+It will also provide configuration to allow for a more "CRD" experience, adjusting the configuration and rules to cater to the needs of CRD authors.
+
+The linter will focus on static evaluation of CRD types.
+It cannot, and will not detect changes to the code, and therefore cannot assert any rule that requires information about a previous version of the types.
+
+### CRD Schema Change Validation
+
+Create a CLI tool that evaluates the differences between an old and new CRD in YAML to identify changes that may break users.
```

Comment:
nit: JSON too?  Or maybe just drop the serialization choice here in KEP and make sure the implementation supports everything needed.